### PR TITLE
feat: Delete proposal

### PR DESCRIPTION
**Pull request by samuveth**
Fixes https://github.com/balancer-labs/snapshot/issues/214

Changes proposed in this pull request:
- When message is `delete-proposal` change the proposal type to `archive-proposal`.
- We might need to increase the `type VARCHAR(12) NOT NULL,` to fit it.
-

### CODE REVIEW

File: server/helpers/adapters/mysql.ts [hunk 1]
```diff
@@ -4,40 +4,54 @@ import fleek from '@fleekhq/fleek-storage-js';
...
+      sig: body.sig,
+      metadata: JSON.stringify({
+        relayer_ipfs_hash: relayerIpfsHash
+      })
+    }
+  ]);
+}
+
+export async function archiveProposal(authorIpfsHash) {
+  let query = 'UPDATE messages SET type = ? WHERE id = ?';
```
Comment (bonustrack) [inline server/helpers/adapters/mysql.ts:33]: It's better to add a LIMIT here
```suggestion
  let query = 'UPDATE messages SET type = ? WHERE id = ? LIMIT 1';
```
Also it's better to do a WHERE type = 'proposal'

File: server/index.ts [hunk 2]
```diff
@@ -106,50 +114,78 @@ router.post('/message', async (req, res) => {
...
+  if (
+    !(await verifySignature(
+      body.address,
+      body.sig,
+      hashPersonalMessage(body.msg)
+    ))
+  )
     return sendError(res, 'wrong signature');
 
+  if (msg.type === 'delete-proposal') {
```
Comment (bonustrack) [inline server/index.ts:148]: Can you move this verification to here https://github.com/snapshot-labs/snapshot-hub/pull/35/files#diff-786114c6ebf98e577ec30dac884800bb7c9ce800b5323710ed03b916e1d33b68R240 ?

File: server/index.ts [hunk 3]
```diff
@@ -106,50 +114,78 @@ router.post('/message', async (req, res) => {
...
+    let query = 'SELECT address FROM messages WHERE id = ?';
```
Comment (bonustrack) [inline server/index.ts:149]: You should explicitly check for type = 'proposal', otherwise it could be used to delete vote

Comment (samuveth) [reply server/index.ts:149]: So the query can be changes by the user?

Comment (bonustrack) [reply server/index.ts:149]: Someone could build a signed message to delete a vote of himself. We dont want this, when a proposal ended the results should not change.

Comment (samuveth) [reply server/index.ts]: Like this e0423ba?

Comment (bonustrack) [reply server/index.ts]: Yes

File: server/index.ts [hunk 4]
```diff
@@ -194,6 +227,17 @@ router.post('/message', async (req, res) => {
     version: '2'
   });
 
+  if (msg.type === 'delete-proposal') {
+    let query = `SELECT address FROM messages WHERE type = 'proposal' AND id = ?`;
+    let propasalSigner = await db.queryAsync(query, [msg.payload.proposal]);
+    if (
+      propasalSigner[0].address.toLowerCase() !== body.address.toLowerCase()
```
Comment (bonustrack) [inline server/index.ts:234]: The address is never lowercase, you could update the query to use `WHERE address = ?` to tackle this

Comment (samuveth) [reply server/index.ts:234]: I couldn't get that to work with `WHERE address = ?` it gives me undefined error that I can't manage to fix. I've removed the `toLowerCase()` now, I'm not sure anymore why I needed them before.

Comment (bonustrack) [reply server/index.ts]: Hum that's weird

File: server/index.ts [hunk 5]
```diff
@@ -194,6 +227,15 @@ router.post('/message', async (req, res) => {
     version: '2'
   });
 
+  if (msg.type === 'delete-proposal') {
+    let query = `SELECT address FROM messages WHERE type = 'proposal' AND id = ?`;
+    let propasalSigner = await db.queryAsync(query, [msg.payload.proposal]);
+    if (propasalSigner[0].address !== body.address) {
+      return sendError(res, 'wrong signer');
+    }
+    await archiveProposal(msg.payload.proposal);
```
Comment (bonustrack) [inline server/index.ts:236]: Actually i mistaken in my previous comment, you was right, the `SELECT` query to check if the action is valid should happen before the `pinJson` function and the `archiveProposal` should happen after, so we need `if (msg.type === 'delete-proposal') {` in 2 different place. Sorry for the hassle!

**Review by bonustrack — APPROVED** [review]
Looking good

**Pull request merged**